File: cli/internal/project/project.go
package project

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const (
	ConfigFile  = ".pb_config.json"
	VersionFile = ".pb_version"
)

type Environment struct {
	InstanceName      string `json:"instanceName,omitempty"`
	TenantID          string `json:"tenantId,omitempty"`
	HealthcheckURL    string `json:"healthcheckBaseUrl,omitempty"`
	HealthcheckMarker string `json:"healthcheckMarker,omitempty"`
}

type Config struct {
	ProjectName string `json:"projectName,omitempty"`
	Pockethost  struct {
		BranchEnvironmentMap map[string]string      `json:"branchEnvironmentMap,omitempty"`
		HealthcheckBaseURL   string                 `json:"healthcheckBaseUrl,omitempty"`
		Environments         map[string]Environment `json:"environments,omitempty"`
	} `json:"pockethost,omitempty"`
}

type Project struct {
	Root    string
	Config  Config
	Version string
}

func FindRoot(start string) (string, error) {
	current, err := filepath.Abs(start)
	if err != nil {
		return "", err
	}
	for {
		configExists := fileExists(filepath.Join(current, ConfigFile))
		versionExists := fileExists(filepath.Join(current, VersionFile))
		if configExists || versionExists {
			return current, nil
		}
		parent := filepath.Dir(current)
		if parent == current {
			return "", fmt.Errorf("could not find %s or %s from %s", ConfigFile, VersionFile, start)
		}
		current = parent
	}
}

func Load(root string) (Project, error) {
	configPath := filepath.Join(root, ConfigFile)
	versionPath := filepath.Join(root, VersionFile)
	if !fileExists(configPath) {
		return Project{}, fmt.Errorf("missing %s in %s", ConfigFile, root)
	}
	if !fileExists(versionPath) {
		return Project{}, fmt.Errorf("missing %s in %s", VersionFile, root)
	}

	configData, err := os.ReadFile(configPath)
	if err != nil {
		return Project{}, err
	}
	var config Config
	if err := json.Unmarshal(configData, &config); err != nil {
		return Project{}, fmt.Errorf("parse %s: %w", configPath, err)
	}
	versionData, err := os.ReadFile(versionPath)
	if err != nil {
		return Project{}, err
	}
	version := strings.TrimSpace(string(versionData))
	if version == "" {
		return Project{}, errors.New(".pb_version is empty")
	}
	return Project{Root: root, Config: config, Version: version}, nil
}

func LoadFrom(start string) (Project, error) {
	root, err := FindRoot(start)
	if err != nil {
		return Project{}, err
	}
	return Load(root)
}

func (p Project) Environment(name string) Environment {
	if p.Config.Pockethost.Environments == nil {
		return Environment{}
	}
	return p.Config.Pockethost.Environments[name]
}

func (p Project) ResolveEnvironment(explicit string) string {
	if explicit != "" {
		return explicit
	}
	if value := os.Getenv("POCKETHOST_ENVIRONMENT"); value != "" {
		return value
	}
	branch := os.Getenv("GITHUB_REF_NAME")
	if branch == "" {
		branch = currentBranch(p.Root)
	}
	if mapped := p.Config.Pockethost.BranchEnvironmentMap[branch]; mapped != "" {
		return mapped
	}
	if branch == "staging" {
		return "staging"
	}
	return "production"
}

func (p Project) ResolveInstance(environment string) string {
	if value := os.Getenv("POCKETHOST_INSTANCE_NAME"); value != "" {
		return value
	}
	configured := p.Environment(environment)
	if configured.InstanceName != "" {
		return configured.InstanceName
	}
	if configured.TenantID != "" {
		return configured.TenantID
	}
	return ""
}

func (p Project) ResolveHealthcheck(environment, instance string) (string, string) {
	if value := os.Getenv("HEALTHCHECK_BASE_URL"); value != "" {
		return value, p.Environment(environment).HealthcheckMarker
	}
	configured := p.Environment(environment)
	if configured.HealthcheckURL != "" {
		return configured.HealthcheckURL, configured.HealthcheckMarker
	}
	if p.Config.Pockethost.HealthcheckBaseURL != "" {
		return p.Config.Pockethost.HealthcheckBaseURL, configured.HealthcheckMarker
	}
	if instance != "" {
		return "https://" + instance + ".pockethost.io", configured.HealthcheckMarker
	}
	return "", configured.HealthcheckMarker
}

func (p Project) Surface() (map[string]bool, error) {
	surface := map[string]bool{}
	for _, name := range []string{"pb_public", "pb_hooks", "pb_migrations"} {
		info, err := os.Stat(filepath.Join(p.Root, name))
		if err != nil {
			if os.IsNotExist(err) {
				surface[name] = false
				continue
			}
			return nil, err
		}
		surface[name] = info.IsDir()
	}
	return surface, nil
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func currentBranch(root string) string {
	data, err := os.ReadFile(filepath.Join(root, ".git", "HEAD"))
	if err != nil {
		return ""
	}
	line := strings.TrimSpace(string(data))
	if !strings.HasPrefix(line, "ref: ") {
		return ""
	}
	parts := strings.Split(strings.TrimPrefix(line, "ref: "), "/")
	return parts[len(parts)-1]
}

File: packages/pocketbase-pockethost/src/project.js
import fs from "node:fs/promises";
import path from "node:path";
import { CONFIG_FILE, DEFAULT_BRANCH_ENVIRONMENT_MAP, VERSION_FILE } from "./constants.js";
import { CommandError } from "./errors.js";
import { fileExists, readJson, readText, slugify } from "./utils.js";

export async function findProjectRoot(startDir = process.cwd()) {
  let currentDir = path.resolve(startDir);

  while (true) {
    const configPath = path.join(currentDir, CONFIG_FILE);
    const versionPath = path.join(currentDir, VERSION_FILE);

    if (await fileExists(configPath) || await fileExists(versionPath)) {
      return currentDir;
    }

    const parentDir = path.dirname(currentDir);
    if (parentDir === currentDir) {
      throw new CommandError(`Could not find ${CONFIG_FILE} from ${startDir}`);
    }

    currentDir = parentDir;
  }
}

export async function loadProject(projectRoot) {
  const configPath = path.join(projectRoot, CONFIG_FILE);
  const versionPath = path.join(projectRoot, VERSION_FILE);

  if (!(await fileExists(configPath))) {
    throw new CommandError(`Missing ${CONFIG_FILE} in ${projectRoot}`);
  }

  if (!(await fileExists(versionPath))) {
    throw new CommandError(`Missing ${VERSION_FILE} in ${projectRoot}`);
  }

  const config = await readJson(configPath);
  const version = (await readText(versionPath)).trim();

  return {
    projectRoot,
    configPath,
    versionPath,
    config,
    version
  };
}

export function getBranchEnvironmentMap(config) {
  return {
    ...DEFAULT_BRANCH_ENVIRONMENT_MAP,
    ...(config.pockethost?.branchEnvironmentMap || {})
  };
}

export async function detectCurrentBranch(projectRoot) {
  const gitHeadPath = path.join(projectRoot, ".git", "HEAD");

  if (!(await fileExists(gitHeadPath))) {
    return null;
  }

  const head = (await fs.readFile(gitHeadPath, "utf8")).trim();

  if (head.startsWith("ref: ")) {
    return head.split("/").pop() || null;
  }

  return null;
}

export async function resolveEnvironmentName(project, options = {}) {
  if (options.environment) {
    return options.environment;
  }

  if (process.env.POCKETHOST_ENVIRONMENT) {
    return process.env.POCKETHOST_ENVIRONMENT;
  }

  if (process.env.GITHUB_REF_NAME) {
    const mapped = getBranchEnvironmentMap(project.config)[process.env.GITHUB_REF_NAME];
    if (mapped) {
      return mapped;
    }

    if (options.strictBranchMapping === true) {
      throw new CommandError(`Unsupported branch '${process.env.GITHUB_REF_NAME}'. Use main, master, staging, or pass --env explicitly.`);
    }
  }

  const currentBranch = await detectCurrentBranch(project.projectRoot);
  if (currentBranch) {
    const mapped = getBranchEnvironmentMap(project.config)[currentBranch];
    if (mapped) {
      return mapped;
    }

    if (options.strictBranchMapping === true) {
      throw new CommandError(`Unsupported branch '${currentBranch}'. Use main, master, staging, or pass --env explicitly.`);
    }
  }

  return "production";
}

export function resolveEnvironmentConfig(project, environmentName) {
  return project.config.pockethost?.environments?.[environmentName] || {};
}

export function resolveInstanceName(project, environmentName) {
  const config = resolveEnvironmentConfig(project, environmentName);
  return process.env.POCKETHOST_INSTANCE_NAME
    || config.instanceName
    || process.env.POCKETHOST_TENANT_ID
    || config.tenantId
    || "";
}

export const resolveTenantId = resolveInstanceName;

export function resolveHealthcheckBaseUrl(project, environmentName, instanceName) {
  const environmentConfig = resolveEnvironmentConfig(project, environmentName);
  const rootBaseUrl = project.config.pockethost?.healthcheckBaseUrl || "";
  const environmentBaseUrl = environmentConfig.healthcheckBaseUrl || "";

  if (process.env.HEALTHCHECK_BASE_URL) {
    return process.env.HEALTHCHECK_BASE_URL;
  }

  if (environmentBaseUrl) {
    return environmentBaseUrl;
  }

  if (rootBaseUrl) {
    return rootBaseUrl;
  }

  if (instanceName) {
    return `https://${instanceName}.pockethost.io`;
  }

  return "";
}

export function resolveProjectName(targetDir) {
  return slugify(path.basename(path.resolve(targetDir)));
}

export async function detectProjectSurface(projectRoot) {
  const pbPublic = await fileExists(path.join(projectRoot, "pb_public"));
  const pbHooks = await directoryHasMeaningfulFiles(path.join(projectRoot, "pb_hooks"));
  const pbMigrations = await directoryHasMeaningfulFiles(path.join(projectRoot, "pb_migrations"));
  const packageJson = await fileExists(path.join(projectRoot, "package.json"));

  return {
    pbPublic,
    pbHooks,
    pbMigrations,
    packageJson
  };
}

async function directoryHasMeaningfulFiles(dirPath) {
  if (!(await fileExists(dirPath))) {
    return false;
  }

  const entries = await fs.readdir(dirPath, { withFileTypes: true });

  for (const entry of entries) {
    if (entry.name === ".gitkeep" || entry.name === ".DS_Store") {
      continue;
    }

    if (entry.isFile()) {
      return true;
    }

    if (entry.isDirectory()) {
      const hasNestedFiles = await directoryHasMeaningfulFiles(path.join(dirPath, entry.name));
      if (hasNestedFiles) {
        return true;
      }
    }
  }

  return false;
}
